grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy

```diff
@@ -0,0 +1,214 @@
+        this.slingRepository = slingRepository
+        //Ensures the job path leads with a '/'
+        this.jobPath = cleanJobPath(jobPath)
+        //Takes the * delimited paths string, creates a collection of relative paths, and normalizes
+        this.relativeExcludePaths = createExcludePaths(this.jobPath, excludePaths)
+    }
+
+
+    /**
+     * @param paths '*' delimited, expected to be driven from job parameters e.g /foo/bar/blah{@literal *}/foo/bar/meh
+     * @param jobPath the job path, used to make sure we are building well-formed relative paths e.g /foo/bar
+     * @return A collection of the exclude paths that are guaranteed to lead with a '/', e.g [blah, meh].  If null or empty, returns an empty collection
+     * @throws IllegalStateException if any paths are not well-formed to the jobPath
+     */
+    private static Collection<String> createExcludePaths(@Nonnull final String jobPath, @Nullable String paths) {
+        final thisPathsString = paths?.trim()
+        if(!thisPathsString) {
+            return []
+        }
+        final theExcludePaths = thisPathsString.split("\\*") as Collection<String>
+        final jobPathWithSlash = jobPath + '/'
+
+        if (theExcludePaths.any {!it.startsWith(jobPathWithSlash)}) throw new IllegalStateException("Not all exclude paths start with \"${jobPath}\": ${theExcludePaths}")
+
+        return theExcludePaths.collect { it.substring(jobPathWithSlash.length()) }
+    }
+
+    /**
+     * @param jobPath expected to be driven from job parameters
+     * @return a "cleaned" job path, i.e leads with a '/', no trailing '/', and trimmed, e.g "/foo/bar"
+     */
+    private static String cleanJobPath(String jobPath) {
+        String thisJobPath = jobPath.trim()
+        //Add the leading '/' if not present
+        thisJobPath = (thisJobPath[0] != '/') ? "/${thisJobPath}" : thisJobPath
+        //Remove any trailing '/'
+        thisJobPath = (thisJobPath[-1] != '/') ? thisJobPath : thisJobPath[0..-2] as String
+
+        return thisJobPath
+    }
+
+
+    /**
+     * Given the current context in the form of a step contribution, do whatever
+     * is necessary to process this unit inside a transaction. Implementations
+     * return {@link RepeatStatus#FINISHED} if finished. If not they return
+     * {@link RepeatStatus#CONTINUABLE}. On failure throws an exception.
+     *
+     * @param contribution mutable state to be passed back to update the current
+     * step execution
+     * @param chunkContext attributes shared between invocations but not between
+     * restarts
+     * @return an {@link RepeatStatus} indicating whether processing is
+     * continuable.
+     */
+    @Override
+    RepeatStatus execute(StepContribution contribution, ChunkContext chunkContext) throws Exception {
+        final Session session = JcrUtil.getSession(slingRepository, "admin")
```

Comment:
Yes I know using ThreadLocal is kind of a last resort general but there were several things that made us use ThreadLocal instead of anything else.

1. The design was simpler to have 1 session per path and we had to share the incoming input stream between steps anyways (not just `clientJcrNodes`)
2. Originally, I was trying to use Spring Batch's `ExecutionContext` to store the session (and the input stream). But, Spring Batch discourages placing large objects in its ExecutionContext, which is understandable since ExecutionContext is one of the objects persisted in the `Spring Batch's Job Repository`.. which for our case, we have our `JcrJobRepository` layer.
3. Hence, ThreadLocal was the last resort and if done properly, we know that it would always work and as mentioned above, we had to share the `inputStream` amongst the steps anyways.